### PR TITLE
Updated for Issue #18

### PR DESCRIPTION
**Pull request by SelerityMichael**
Add user feedback to startup process.

### CODE REVIEW

File: launchapro.sh [hunk 1]
```diff
@@ -7,14 +7,31 @@
 SCRIPT_ROOT=$(dirname $0)
 cd ${SCRIPT_ROOT}
 
+# Check that docker is running
+DOCKER_STATUS=$(docker version 2>&1 > /dev/null)
```
Comment (danozdotnet) [inline launchapro.sh:11]: as this variable is never used and you only care about the exit code, i would suggest running like so:

# Check that docker is running
if ! docker version 2>&1 >/dev/null; then

File: launchapro.sh [hunk 2]
```diff
@@ -98,11 +115,77 @@ RUN_ARGS="
...
 --volume ${PWD}/python:/python
 --volume ${PWD}/data:/data
 ${JUPYTERLAB_ARGS}"
 
 # Run Analytics Pro container with supplied arguments
-docker run -u root ${RUN_ARGS} "${IMAGE}:${IMAGE_VERSION}" "${@}"
+CONTAINER=$(docker run -u root ${RUN_ARGS} "${IMAGE}:${IMAGE_VERSION}" "${@}")
+
+# Check if there were any problems with the launch
+if [[ $? > 0 ]]; then
```
Comment (danozdotnet) [inline launchapro.sh:128]: the same can be done here if you'd like:

# Run Analytics Pro container with supplied arguments
# and check if there were any problems with the launch
if ! docker run -u root ${RUN_ARGS} "${IMAGE}:${IMAGE_VERSION}" "${@}"; then

Comment (danozdotnet) [reply launchapro.sh:128]: actually I see the variable is used, but you could still run it in a similar manner, or just leave as is :)

if ! CONTAINER=$(docker run -u root ${RUN_ARGS} "${IMAGE}:${IMAGE_VERSION}" "${@}"); then

File: launchapro.sh (hunk 1 above)
Comment (SelerityMichael) [reply launchapro.sh:11]: In its current state if docker isn't running this is the experience:
```
michaeldixon@Michaels-MBP sas-analytics-pro % ./launchapro.sh 
ERROR: A running docker client is required to use this software.  Please install or start your instance of Docker before proceeding.
```

The proposed change does this:
```
michaeldixon@Michaels-MBP sas-analytics-pro % ./launchapro.sh 
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
ERROR: A running docker client is required to use this software.  Please install or start your instance of Docker before proceeding.
```

We want to fully control what is presented to the user, but this gives the output from running the docker command?

File: launchapro.sh (hunk 2 above)
Comment (SelerityMichael) [reply launchapro.sh:128]: Let's leave this as-is unless it gives us problems in the future.

File: launchapro.sh (hunk 1 above)
Comment (danozdotnet) [reply launchapro.sh:11]: the redirects are around the wrong way, haha. It's supposed to be this way ;)
>/dev/null 2>&1

File: launchapro.sh [hunk 3]
```diff
@@ -98,11 +115,77 @@ RUN_ARGS="
...
+  echo "#    SAS Analytics Pro Personal Launcher    #"
+  echo "#-------------------------------------------#"
+  echo "# S = SAS Studio has started                #"
+  if [[ ${JUPYTERLAB} == "true" ]]; then
+    echo "# J = Jupyter Lab has started               #"
+  fi
+  echo "#############################################"
+  echo -n "."
+  TIMING="5 5 5 5 10 10 30 30 30 60"
+  for _check in ${TIMING}; do
```
Comment (danozdotnet) [inline launchapro.sh:153]: one last idea, in this block we run the 'docker logs' command three times in a row, which is in a loop. realistically we should just run it once per loop. by running it separately for each variable, there is the potential that it doesn't have the password yet, then when it checks that studio has started it has, so it breaks out of the loop, leaving the user without a password. very unlikely possibility, but still possible.

so I suggest at the start of the loop, we get the docker log output, then each variable checks that output so it's consistent for each. 

```
  for _check in ${TIMING}; do
    sleep "${_check}"
    DOCKER_LOGS="$(docker logs "${CONTAINER}" 2>&1)"
    APRO_PASSWORD=${APRO_PASSWORD:-$(grep ^Password= <<< "${DOCKER_LOGS}")}
    STUDIO_START=${STUDIO_START:-$(grep "service Root WebApplicationContext: initialization completed" <<< "${DOCKER_LOGS}")}
    
    if [[ ${JUPYTERLAB} == "true" ]]; then
      JUPYTER_START=${JUPYTER_START:-$(grep "Jupyter Server " <<< "${DOCKER_LOGS}")}
    fi
```

File: launchapro.sh (hunk 1 above)
Comment (SelerityMichael) [reply launchapro.sh:11]: Tested solution on MacOS and Ubuntu.

File: launchapro.sh (hunk 3 above)
Comment (SelerityMichael) [reply launchapro.sh:152]: Much better use of resources.

**Review by danozdotnet — APPROVED** [review]

**Pull request merged**